### ISSUE TITLE
Figure out external data sources

### ISSUE THREAD

**Issue by rpearl** [rpearl/chlorophyll#51]
Wneed to figure out the "external environment" interface: specifying the data stream format (we'd talked about OSC as a message option here), and more particularly figuring out how we *discover* sources of data. I can imagine an "internal" source of data with a nice UI we design for doing things like "sweep a line through this rotation", and I can imagine external sources of data for sensors (or maybe you want to program your line's behavior in a specific language). It could be the case that there are dependency chains too: producing a line whose rotation depends on the magnitude of a sensor signal modulated with an internal clock source.

I can imagine the sequencer being implemented as a data stream in precisely the same way as well: messages being interpreted as commands to trigger patterns on groups.

**Comment by cwill** [comment]
Yeah, it seems we're approaching a general workflow of:

1. create {patterns, object animations, ...} with various tools (pattern editor, mapping editor, ...)
2. specifying things like what inputs are parameterized (e.g. rotation angle, color, time value)
3. create inputs, whether artificial/internal (a knob inside chlorophyll for test input), or external (a midi controller)
4. Map inputs to outputs.

Sounds like it'll probably be useful to separate "input tools" in a similar manner to how we've done "output tools" (mappings, patterns).  So we might have a tool to create an internal-data-source as you mentioned (sweep a line, output a position, output this sine wave, trigger an event every 0.5s).  We also probably want something similar for real-world inputs, to allow mapping input from an ipad or midi controller into several discrete outputs, which can then be consumed by {patterns, object animations, sequencers}.

This feels kind of like we want some sort of patch-bay type interface, to allow configuring and plugging all these modules into each other after they've been created. There's a lot of precedent for this sort of thing in audio contexts:

[Reason uses a literal patch bay](http://dt7v1i9vyp3mf.cloudfront.net/styles/news_large/s3/imagelibrary/r/reason2-1004.jpg?euGqveBkA2HHP1YZjm4Uvc9zZ9uWnfqb=&itok=J42ZPMWS) which looks like it turns into a mess really quickly but is pretty intuitive.
[Ableton uses a more typical dropdown-boxes approach](https://cdn2-resources.ableton.com/80bA26cPQ1hEJDFjpUKntxfqdmG3ZykO/static/images/manual/en/SessionMixerInOutSectio_fmt.png) which looks more compact, but makes it harder to tell what's going where at a glance.

I think we might be able to use the new&improved litegraph stuff here? It'd let us give a nice input->output mapping interace which clearly shows all the connections involved, but lets us hide fiddly parameters & such in the config panels for the nodes themselves instead of having it all as graph-programming based on raw inputs, which seems like it'd get tedious and messy.

**Comment by rpearl** [comment]
The patch-bay stuff is interesting but (currently, anyway) all of the "modulation" of the signal comes in the pattern graphs we already have, so a two-column dropdown routing external input channels to particular consumers might be all we need here...

**Comment by cwill** [comment]
Yeah, dropdowns are probably sufficient. If it turns out to get confusing for more complicated pieces we can always add more visual sugar later.

**Comment by cwill** [comment]
We've settled on OSC as a messaging format, so renaming this to focus on tracking making a UI for creating & routing outside inputs and control signals.
